Add prometheus listener documentation on Manager, Horizon and terraform (#5936)
ovh/docs: pages/public_cloud/public_cloud_network_services/technical-resources-02-octavia-monitoring-prometheus/guide.en-gb.md

## Instructions

> ### Pre requisites

Reviewer: ```suggestion
### Prerequisites

```

### Monitoring with Prometheus

> To add a Prometheus endpoint on a Public Cloud Load Balancer, create a listener with a `PROMETHEUS` special protocol. This will enable the  `/metrics` endpoint which by default listens on all the interfaces from your load balancer. **This means that if your load balancer is using a Floating IP, by default the `/metrics` endpoints will be opened to the internet**. To avoid that, we advise to set the `allowed_cidrs`` option to protect your metrics, for example with the CIDR of the subnet where the load balancer is spawned.

Reviewer: ```suggestion
To add a Prometheus endpoint on a Public Cloud Load Balancer, create a listener using the `PROMETHEUS` protocol. This will enable the  `/metrics` endpoint which by default listens on all the interfaces from your load balancer. **This means that if your load balancer is using a Floating IP, by default the `/metrics` endpoint will be opened to the internet**. To avoid that, we advise to set the `allowed_cidrs` option to protect your metrics, for example with the CIDR of the subnet where the load balancer is spawned.
```

This listener type provides the same features as the "regular" listeners but does not support attaching pools or L7 policies. All metrics will be identified by the Octavia object ID (UUID) of the resources.

> [!primary]
>
> Currently UDP and SCTP specific metrics are not reported via Prometheus endpoints.
>

> #### Create a Prometheus listener via OVHcloud Control Panel

Reviewer: ```suggestion
#### Create a Prometheus listener via OVHcloud Control Panel

```

Under `Network` category, select `Load Balancer`. 
A page listing the load balancer will be displayed. 
Select one load balancer by clicking on its `name`.
The load balancer details page will be displayed. Click on the `Listeners` tab then click on `Add a listener`

In the form:
* fill in the `Name`
* select `prometheus` in the Protocol

> * choose a port (different from the already existing listener ports of your Load Balancer)

Reviewer: ```suggestion
* choose a port (different from the already existing listener ports of your Load Balancer)

```

> The default pool is greyed out because the `prometheus` listener does not need a pool contrary to the other types of listeners.

Reviewer: ```suggestion
The default pool is greyed out because the `prometheus` listener does not need a pool unlike the other types of listeners.
```

![Create listener in OVHCloud Control Panel](img/create_listener_manager.png)

Click on `Add` to trigger the listener creation. 
The new listener is added in the listener list.

![Listener list in OVHCloud Control Panel](img/listener_list.png)

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
```

#### Create a Prometheus listener via Openstack GUI (Horizon)

Log in to Horizon by following this [guide](../../compute/introducing_horizon/)
Click on `Network` > `Load Balancers`
The Load Balancer list is displayed. 

> Click on the load balancer name. The load balancer details page is displayed.

Reviewer: ```suggestion
Click on the load balancer name. The load balancer details page is displayed.

```

![Load Balancer details in Horizon](img/horizon_lb_details.png)

Click on the `Listeners` tab then `Create Listener`

> In the listener creation page, fill in the `Name` and the `Protocol` to `PROMETHEUS`. The port will be set to a default value:  change it if needed.

Reviewer: ```suggestion
In the listener creation page, fill in the `Name` and the `Protocol` to `PROMETHEUS`.
The port will be set to a default value, change it if needed.
```

![Listener creation in listener](img/horizon_listener_creation.png)

> 

Reviewer: ```suggestion
```

+-----------------------------+--------------------------------------+
```

> Note that you add the `--allowed-cidr` option in order to filter the listener to a specific network.

Reviewer 2: Maybe we should highlight that the metric listener will be exposed on Internet (if a FIP is attached) ?

Author: Thanks @Lucasgranet  I have modified the introduction to specify that.

+-----------------------------+--------------------------------------+
```

Note that you should add the `--allowed-cidr` option in order to filter the listener to a specific network.

> #### Create a Prometheus listener via terraform

Reviewer: ```suggestion
#### Create a Prometheus listener via Terraform

```

> The ressource [openstack_lb_listener_v2](https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs/resources/lb_listener_v2) from the Openstack provider enables to configure a prometheus listener. The following snippet is extracted from a full example available on [github](https://github.com/yomovh/tf-at-ovhcloud/tree/main/simple_http_lb_with_prom_grafana), adapt it to your need. 

Reviewer: ```suggestion
The resource [openstack_lb_listener_v2](https://registry.terraform.io/providers/terraform-provider-openstack/openstack/latest/docs/resources/lb_listener_v2) from the Openstack provider enables to configure a Prometheus listener. The following snippet is extracted from a full example available on [Github](https://github.com/yomovh/tf-at-ovhcloud/tree/main/simple_http_lb_with_prom_grafana), adapt it to your need. 
```

> 

Reviewer: ```suggestion
```

```hcl
resource "openstack_lb_listener_v2" "prom_listener" {
  protocol        = "PROMETHEUS"
  protocol_port   = 9100
  loadbalancer_id = openstack_lb_loadbalancer_v2.tf_lb.id
  #restrict the access of the listener to the private network subnet
  allowed_cidrs = [openstack_networking_subnet_v2.tf_lb_subnet.cidr]
}

> ```

Reviewer: ```suggestion
```

```

> #### Configure Prometheus to collect your metrics

Reviewer: ```suggestion
#### Configure Prometheus to collect your metrics

```

Once the `PROMETHEUS` listener is `ACTIVE`, you can configure Prometheus to collect metrics from the Load Balancer by updating the `prometheus.yml` file.

```yaml
[scrape_configs]
- job_name: 'Octavia LB1'
  static_configs:
  - targets: ['192.0.2.10:9100']
```

For more information on setting up Prometheus, see the [Prometheus project website](https://prometheus.io/).

You can connect [Grafana](https://grafana.com) to the [Prometheus](https://prometheus.io) instance to provide additional graphing and dashboard capabilities. A Grafana dashboard for Octavia Load Balancers is available [here](https://grafana.com/grafana/dashboards/15828-octavia-amphora-load-balancer/)

> The metrics have the following naming format :

Reviewer: ```suggestion
The metrics have the following naming format:
```

|octavia_member_client_aborts_total|Total number of data transfers aborted by the client.|
|octavia_member_server_aborts_total|Total number of data transfers aborted by the server.|
|octavia_member_weight|Member weight.|
|octavia_member_check_failures_total|Total number of failed check (Only counts checks failed when the member is up).|
|octavia_member_check_up_down_total|Total number of UP->DOWN transitions.|
|octavia_member_downtime_seconds_total|Total downtime (in seconds) for the member.|
|octavia_member_check_last_change_seconds|Number of seconds since the last UP<->DOWN transition.|
|octavia_member_current_throttle|Current throttle percentage for the member, when slowstart is active, or no value if not in slowstart.|
|octavia_member_loadbalanced_total|Total number of times a member was selected, either for new sessions, or when redispatching.|
|octavia_member_http_responses_total|Total number of HTTP responses.|
|octavia_member_idle_connections_current|Current number of idle connections available for reuse|
|octavia_member_idle_connections_limit|Limit on the number of available idle connections|

> 

Reviewer: ```suggestion
```
